iio: dac: Add support for AD5676R/AD5696R DACs

stefpopa opened the pull request:
This patch adds support for AD5676R (spi) and AD5696R (i2c)
family of DAC devices.

Signed-off-by: Stefan Popa <stefan.popa@analog.com>

lclausen-adi commented on drivers/iio/dac/ad5686.c:
```diff
@@ -384,23 +447,210 @@ static int ad5686_remove(struct spi_device *spi)
+	val <<= shift;
+	msg = cpu_to_be32((AD5686_CMD(cmd) | AD5686_ADDR(addr) | val) << 8);
+
+	ret = i2c_master_send(i2c, (char *)&msg, 3);
```
Try to use the transfer buffer instead of a on-stack variable, just like the SPI read write functions.

on-stack buffers are not always safe for I2C transfers in case for example the I2C driver tries to do DMA.

commodo replied:
if doing the change to a buffer, `i2c_master_send(...., ...., 3);`  to ``i2c_master_send(...., ...., len);`
and init `len = sizeof(buf)` ; that way you can also check if `ret == len`

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -304,25 +329,68 @@ static const struct ad5686_chip_info ad5686_chip_info_tbl[] = {
 
-
-static int ad5686_probe(struct spi_device *spi)
+static int ad5686_probe(struct device *dev, enum ad5686_supported_device_ids chip_type,
```
i'd rename this to `ad5686_probe_common()` to make sure it's not assumed to be a probe function on it's own

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -335,31 +403,26 @@ static int ad5686_probe(struct spi_device *spi)
+	st->chip_info = &ad5686_chip_info_tbl[chip_type];
 
-	if (voltage_uv)
+	/* If available, internal reference is on by default */
```
this change can be it's own commit [ changing to the internal reference by default ]

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -335,31 +403,26 @@ static int ad5686_probe(struct spi_device *spi)
 	indio_dev->channels = st->chip_info->channel;
 	indio_dev->num_channels = AD5686_DAC_CHANNELS;
 
-	ret = ad5686_spi_write(st, AD5686_CMD_INTERNAL_REFER_SETUP, 0,
```
curios: this change removes this `write` ; is this fine to just remove it ? or does this need to be replaced with a `st->write()` equivalent ?

stefpopa replied:
The on-chip reference is on at power-up by default. The driver uses the internal reference if it is supported. In this case it is not compulsory to do a write.

commodo replied:
ack ; thanks for clarification

lclausen-adi replied:
so, what happens if somebody wants to use an external reference? Previously this was possible with this driver, now it no longer is.

stefpopa replied:
You are right. I will change this.

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -384,23 +447,210 @@ static int ad5686_remove(struct spi_device *spi)
+static const struct i2c_device_id ad5686_i2c_id[] = {
+	{"ad5671r", ID_AD5671R},
+	{"ad5675r", ID_AD5675R},
+	{"ad5694", ID_AD5694},
```
nitpick: align the IDs column

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -384,23 +447,210 @@ static int ad5686_remove(struct spi_device *spi)
+	char data[3];
+	int ret;
+
+	ret = i2c_master_recv(i2c, data, 2);
```
this reads 2 bytes, but data is 3 bytes ; 
if 2 bytes are needed i'd reduce `char data[2]` 

either way, i'd use `sizeof(data)` here

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -384,23 +447,210 @@ static int ad5686_remove(struct spi_device *spi)
+
+	ret = i2c_master_send(i2c, (char *)&msg, 3);
+
+	return (ret < 0 ? ret : (ret != 3) ? -EIO : 0);
```
this is usually not preferred when doing validation, as it's a bit hard to follow in some points of the day [after the 2nd half of the day usually]
i'd split this into a simpler form ; something like:

```
if (ret < 0)
   return ret;

return (ret != len) ? -EIO : 0;
```

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -384,23 +447,210 @@ static int ad5686_remove(struct spi_device *spi)
+	{"ad5684r", ID_AD5684R},
+	{"ad5685r", ID_AD5685R},
 	{"ad5686", ID_AD5686},
+	{"ad5686r", ID_AD5686R},
```
nitpick: align the IDs column

lclausen-adi replied:
I would not do that, it means you have to re-indent the whole thing when you add a new ID with a different length. That causes a lot of noise in the changes.

commodo replied:
fine from me to not change :)

stefpopa commented:
Res-pinned based on review comments

commodo commented on drivers/iio/dac/ad5686-spi.c:
```diff
@@ -0,0 +1,94 @@
+/*
```
ideally there should be a commit that deals with just the separation of the spi part into it's own file ;
then in a subsequent commit, the new chips can be added

stefpopa replied:
Thought about it, but the spi part separation makes sense only in the context of adding support for a family of devices which use the i2c interface.

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -284,31 +289,72 @@ static const struct iio_chan_spec_ext_info ad5686_ext_info[] = {
 	},
 };
 
-
```
nitpick: this is whitespace change ; can go in first commit :)

lclausen-adi commented on drivers/iio/dac/Kconfig:
```diff
@@ -129,11 +129,27 @@ config AD5624R_SPI
+	  module will be called ad5686.
+
+config AD5686_I2C
+	tristate "Analog Devices AD5686 and similar multi-channel DACs"
```
We now have two Kconfig entries with the same label. I'd add (SPI) and (I2C) at the end. Maybe also change the part number for the I2C entry, since the AD5686 is SPI only.

lclausen-adi commented on drivers/iio/dac/ad5686-i2c.c:
```diff
@@ -0,0 +1,87 @@
+	ret = i2c_master_recv(i2c, &st->data[0].d8[1], 2);
+	if (ret < 0)
+		return ret;
+
```
Can you check whether it works when you combine the i2c_master_send and i2c_master_recv into one i2c_transfer transaction with a read and write part? Splitting this into two separate transactions has the disadvantage that the I2C bus is released in between and another I2C driver might schedule a transaction on the same bus in between the two. This can cause all kind of difficult to diagnose issues since this would be a very rare race condition

lclausen-adi commented on drivers/iio/dac/ad5686-i2c.c:
```diff
@@ -0,0 +1,87 @@
+		return ret;
+
+	return (((int)st->data[0].d8[1] << 8) | st->data[0].d8[2]);
+}
```
You could add a __be16 d16 field to the union and then use be16_to_cpu() here. And also pass the pointer to the d16 field as the I2C read buffer.

lclausen-adi commented on drivers/iio/dac/ad5686.c:
```diff
@@ -193,8 +78,8 @@ static ssize_t ad5686_write_dac_powerdown(struct iio_dev *indio_dev,
 
-	ret = ad5686_spi_write(st, AD5686_CMD_POWERDOWN_DAC, 0,
-			       st->pwr_down_mask & st->pwr_down_mode, 0);
+	ret = st->write(st, AD5686_CMD_POWERDOWN_DAC, 0,
```
I'd wrap the st->write and st->read access in helper functions ad5686_write() and ad5686_read(). This provides a clean separation between implementation and function.

stefpopa commented:
changelog v2 -> v3:
- Changed the part number for the I2C entry from AD5686_I2C to AD5696_I2C and modified the label.
- used one i2c_transfer transaction and added a _be16 d16 field to the union which is passed as a pointer to the I2C read buffer
- Wrapped the st->write and st->read access in helper functions ad5686_write() and ad5686_read()

lclausen-adi commented:
Looks good. But I think the last two patches got squashed together?

commodo commented:
[![LGTM](https://i.pinimg.com/474x/cf/e2/07/cfe2073d31d2eaa873890caec825486f--giant-minion-evil-minions.jpg)](https://i.pinimg.com/474x/cf/e2/07/cfe2073d31d2eaa873890caec825486f--giant-minion-evil-minions.jpg)
